File: 프로그래머스/2/42587. 프로세스/프로세스.py
def solution(priorities, location):
    q = []
    answer = 1
    for i in range(len(priorities)):
        q.append((priorities[i], i))
    
    while q:
        now = q[0][0]
        idx = q[0][1]
        for val, cnt in q:
            if now < val:
                q.append((now, idx))
                q.pop(0)
                flag = False
                break
        else:
            if (idx == location):
                break
            else:
                q.pop(0)
                answer+=1
        
    
    
    return answer

File: 프로그래머스/2/42587. 프로세스/프로세스.java
import java.util.*;

class Solution {
    public int solution(int[] priorities, int location) {
        int answer = 0;
        
        Map<Integer, Integer> map = new HashMap<>();
        Queue<Integer> q = new LinkedList<>();
        PriorityQueue<Integer> pq = new PriorityQueue<>( (a,b) -> map.get(b) - map.get(a) );
        
        for (int i=0;i<priorities.length;i++) {
            map.put(i, priorities[i]);
            q.add(i);
            pq.add(i);
        }
        
        while(!q.isEmpty()) {
            int popNum = q.poll();
            int target = pq.peek();
            
            if (map.get(popNum) == map.get(target)) {
                pq.poll();
                answer++;
                
                if (popNum == location) return answer;
            }
            else {
                q.add(popNum);
            }
            
            
        }
        
        
        
        return answer;
    }
}